Cython/Compiler/ExprNodes.py

```diff
@@ -8779,6 +8780,86 @@ def generate_evaluation_code(self, code):
             arg.free_temps(code)
 
 
+class FrozenSetNode(SetNode):
+    """
+    Frozenset constructor.
+    """
+
+    gil_message = "Constructing Python frozenset"
+    result_code = None
+    is_literal = False
+
+    def calculate_result_code(self):
+        return self.result_code
+
+    def _generate_creating_from_constant_code(self, code, target):
+        if target is None:
+            target = self.result()
+        assert target is not None
+        if self.args is None:
+            constructor_parameter = 0
+        else:
+            self.args.generate_evaluation_code(code)
+            constructor_parameter = self.args.py_result()
+        code.putln("%s = PyFrozenSet_New(%s); %s" % (
+            target, constructor_parameter,
+            code.error_goto_if_null(target, self.pos)))
+        self.result_code = target
+
+# TODO: for such case
+#     x = frozenset((1, 2, 3, 2))
+#     y = frozenset((1, 2, 3, 2, 3, 1))
+#     _get_dedup_values will generate same dedup key,
+#     but cython will still try to create frozenset from a tuple with duplicate elements
+
+    def _get_dedup_values(self, args):
+        result = []
+        result_set = set()
+        for argument in args:
+            val = argument.constant_result
+            if val is None:
+                return args  # We expect a constant value. If not, we skip this phase
```

Comment:
`None` is a valid constant. Please add a test that includes `None` as frozenset item.
```suggestion
            if not argument.has_constant_result():
                return args  # Constant frozensets require constant items.
            val = argument.constant_result
```

Reply:
I don't think so. If we have a` None` in frozenset, like `a = frozenset((None,1))` , we expect a valid `NoneNode`
```
<Cython.Compiler.ExprNodes.NoneNode object at 0x7f38e5cd7910>
{'pos': (<FileSourceDescriptor:/home/lizhenbo/src/cython/sno.pyx>, 3, 15)}
```

Reply:
Yes, and the `constant_result` of such a `NoneNode` is `None`. That's not the case you want to handle here.